File: companion/lib/discover.js
const fs = require('fs');
const os = require('os');
const path = require('path');

// 在一组候选 transcript 里选:mtime≥startedAt(含 5s 宽限)且 sid 未被占用的最新一个
function pickLatestTranscript(files, startedAtMs, claimedSids) {
  const grace = startedAtMs - 5000;
  return files
    .filter((f) => f.mtimeMs >= grace && !claimedSids.has(f.sid))
    .sort((a, b) => b.mtimeMs - a.mtimeMs)[0] || null;
}

function listTranscripts(projDir) {
  try {
    return fs.readdirSync(projDir)
      .filter((n) => n.endsWith('.jsonl'))
      .map((n) => ({ sid: n.replace(/\.jsonl$/, ''), path: path.join(projDir, n), mtimeMs: fs.statSync(path.join(projDir, n)).mtimeMs }));
  } catch (_) { return []; }
}

// cwd -> Claude project 目录名(把非字母数字换成 '-')
function encodeCwd(cwd) {
  return cwd.replace(/[^a-zA-Z0-9]/g, '-');
}

function countSubagents(subagentsDir) {
  try { return fs.readdirSync(subagentsDir).filter((n) => n.endsWith('.meta.json')).length; }
  catch (_) { return 0; }
}

// homeDirs: 形如 [~/.claude, ~/.claude-work];aliverPids: 存活 pid 集合
function discoverClaudeSessions({ homeDirs, aliverPids }) {
  const out = [];
  const claimed = new Set();
  for (const claude of homeDirs) {
    const sessDir = path.join(claude, 'sessions');
    let files = [];
    try { files = fs.readdirSync(sessDir).filter((n) => n.endsWith('.json')); } catch (_) { continue; }
    for (const fname of files) {
      let meta;
      try { meta = JSON.parse(fs.readFileSync(path.join(sessDir, fname), 'utf8')); } catch (_) { continue; }
      if (!meta.pid || !aliverPids.has(meta.pid)) continue;
      const projDir = path.join(claude, 'projects', encodeCwd(meta.cwd || ''));
      const transcripts = listTranscripts(projDir);
      const direct = transcripts.find((t) => t.sid === meta.sessionId);
      const chosen = direct || pickLatestTranscript(transcripts, meta.startedAt || 0, claimed);
      if (!chosen) continue;
      claimed.add(chosen.sid);
      out.push({
        configRoot: claude, pid: meta.pid, sessionId: chosen.sid, cwd: meta.cwd || '',
        startedAt: meta.startedAt || 0, transcriptPath: chosen.path,
        subagentsDir: path.join(projDir, chosen.sid, 'subagents'),
      });
    }
  }
  return out;
}

// 默认 Claude 配置目录:~/.claude* + $CLAUDE_CONFIG_DIR
function defaultClaudeHomeDirs() {
  const home = os.homedir();
  const dirs = new Set();
  if (process.env.CLAUDE_CONFIG_DIR) dirs.add(process.env.CLAUDE_CONFIG_DIR);
  try {
    for (const n of fs.readdirSync(home)) if (/^\.claude/.test(n)) {
      const p = path.join(home, n);
      if (fs.existsSync(path.join(p, 'sessions'))) dirs.add(p);
    }
  } catch (_) {}
  if (dirs.size === 0) dirs.add(path.join(home, '.claude'));
  return [...dirs];
}

// 扫 ~/.codex/sessions/**/rollout-*.jsonl,返回最近 N 个(按 mtime)
function listCodexRollouts(codexRoot, limit = 40) {
  const root = path.join(codexRoot, 'sessions');
  const out = [];
  const walk = (dir) => {
    let ents = [];
    try { ents = fs.readdirSync(dir, { withFileTypes: true }); } catch (_) { return; }
    for (const e of ents) {
      const p = path.join(dir, e.name);
      if (e.isDirectory()) walk(p);
      else if (/^rollout-.*\.jsonl$/.test(e.name)) { try { out.push({ path: p, mtimeMs: fs.statSync(p).mtimeMs }); } catch (_) {} }
    }
  };
  walk(root);
  return out.sort((a, b) => b.mtimeMs - a.mtimeMs).slice(0, limit);
}

module.exports = {
  pickLatestTranscript, listTranscripts, encodeCwd, countSubagents,
  discoverClaudeSessions, defaultClaudeHomeDirs, listCodexRollouts,
};

File: companion/codey/discover.py
"""文件系统发现会话(读 ~/.claude*/sessions、transcript、~/.codex/sessions)。"""
import json
import os
import re


def pick_latest_transcript(files, started_at_ms, claimed_sids):
    """选 mtime>=startedAt(含 5s 宽限)且 sid 未占用的最新 jsonl。"""
    grace = started_at_ms - 5000
    cands = [f for f in files if f["mtime_ms"] >= grace and f["sid"] not in claimed_sids]
    cands.sort(key=lambda f: f["mtime_ms"], reverse=True)
    return cands[0] if cands else None


def list_transcripts(proj_dir):
    try:
        names = [n for n in os.listdir(proj_dir) if n.endswith(".jsonl")]
    except OSError:
        return []
    out = []
    for n in names:
        p = os.path.join(proj_dir, n)
        try:
            out.append({"sid": n[:-6], "path": p, "mtime_ms": os.stat(p).st_mtime * 1000})
        except OSError:
            continue
    return out


def encode_cwd(cwd):
    return re.sub(r"[^a-zA-Z0-9]", "-", cwd)


def count_subagents(subagents_dir):
    try:
        return sum(1 for n in os.listdir(subagents_dir) if n.endswith(".meta.json"))
    except OSError:
        return 0


def discover_claude_sessions(home_dirs, aliver_pids):
    out, claimed = [], set()
    for claude in home_dirs:
        sess_dir = os.path.join(claude, "sessions")
        try:
            files = [n for n in os.listdir(sess_dir) if n.endswith(".json")]
        except OSError:
            continue
        for fname in files:
            try:
                with open(os.path.join(sess_dir, fname), encoding="utf-8") as fh:
                    meta = json.load(fh)
            except (OSError, ValueError):
                continue
            pid = meta.get("pid")
            if not pid or pid not in aliver_pids:
                continue
            proj_dir = os.path.join(claude, "projects", encode_cwd(meta.get("cwd") or ""))
            transcripts = list_transcripts(proj_dir)
            direct = next((t for t in transcripts if t["sid"] == meta.get("sessionId")), None)
            chosen = direct or pick_latest_transcript(transcripts, meta.get("startedAt") or 0, claimed)
            if not chosen:
                continue
            claimed.add(chosen["sid"])
            out.append({
                "config_root": claude, "pid": pid, "session_id": chosen["sid"],
                "cwd": meta.get("cwd") or "", "started_at": meta.get("startedAt") or 0,
                "transcript_path": chosen["path"],
                "subagents_dir": os.path.join(proj_dir, chosen["sid"], "subagents"),
            })
    return out


def default_claude_home_dirs():
    home = os.path.expanduser("~")
    dirs = []
    env = os.environ.get("CLAUDE_CONFIG_DIR")
    if env:
        dirs.append(env)
    try:
        for n in os.listdir(home):
            if n.startswith(".claude"):
                p = os.path.join(home, n)
                try:
                    if os.path.isdir(os.path.join(p, "sessions")) and p not in dirs:
                        dirs.append(p)
                except OSError:
                    pass
    except OSError:
        pass
    if not dirs:
        dirs.append(os.path.join(home, ".claude"))
    return dirs


def list_codex_rollouts(codex_root, limit=40):
    root = os.path.join(codex_root, "sessions")
    out = []

    def walk(d, depth=0):
        if depth > 5:
            return
        try:
            entries = list(os.scandir(d))
        except OSError:
            return
        for e in entries:
            try:
                if e.is_symlink():
                    continue
                if e.is_dir():
                    walk(e.path, depth + 1)
                elif re.match(r"^rollout-.*\.jsonl$", e.name):
                    out.append({"path": e.path, "mtime_ms": e.stat().st_mtime * 1000})
            except OSError:
                continue

    walk(root, 0)
    out.sort(key=lambda x: x["mtime_ms"], reverse=True)
    return out[:limit]
